controllers/installation_controller.go

```diff
@@ -619,9 +424,17 @@ func (r *InstallationReconciler) ReconcileK0sVersion(ctx context.Context, in *v1
 	}
 
 	if in.Spec.AirGap {
+		image := meta.Artifacts["local-artifact-mirror-image"]
+		if image == "" {
+			reason := "No local-artifact-mirror-image defined in release bundle"
+			in.Status.SetState(v1beta1.InstallationStateFailed, reason, nil)
+			return nil
+		}
+		log.Info("Using local-artifact-mirror image", "image", image)
+
 		// in airgap installations let's make sure all assets have been copied to nodes.
 		// this may take some time so we only move forward when 'ready'.
-		if ready, err := r.CopyArtifactsToNodes(ctx, in); err != nil {
+		if ready, err := r.CopyArtifactsToNodes(ctx, in, image); err != nil {
```

Comment:
Wouldn't this be called twice during an upgrade now? Once by the operator upgrade command, and another during reconciliation? Artifacts should be distributed at this point, and I don't see a scenario where this still applies.

Reply:
Yes it will, although my hope is that it would do nothing if it were called a second time. Unfortunately it is really hard for me to test this end to end as i have to apply manual patches to get the upgrade to start with the images and binaries that I am using and once the cluster is upgraded it reverts back to the old assets built by the [EC CI](https://github.com/replicatedhq/embedded-cluster/pull/684#issuecomment-2159134964).

Reply:
I understand. Even if it's going to be a no-op, I think this logic shouldn't be executed here anymore. It made sense in the old workflow, but it doesn't with the new one. It's unnecessary code complexity.

Reply:
removed

Reply:
I was thinking more about this. I think this is still needed for upgrades prior to code with the new upgrade logic. If you agree I will add it back.

Reply:
I don't agree. New upgrade logic will always use the new code. Old upgrade logic will always use the old code. Everything is coupled.

When upgrading from old code:

- Old KOTS creates an installation object
- Old operator distributes artifacts
- Old operator upgrades the cluster

Reply:
i think this does belong here as well since a controllers job is to reconcile the desired state against the actual state. this will make it a more resilient system. unless the operator is not responsible at all for distributing artifacts and that is another components responsibility. if so i could see removing this.

Reply:
That may be the case for resources deployed to the cluster, but these artifacts are only used during the upgrade process. It's also an expensive process specially when there are a large number of nodes. I'm also worried that this will basically always run twice during an upgrade with the new upgrade workflow. Isn't that the case? If so, that seems wrong.

Reply:
I removed it